### PR TITLE
8248862: Implement Enhanced Pseudo-Random Number Generators

### PR DESCRIPTION
**Pull request by JimLaskey**
This PR is to introduce a new random number API for the JDK. The primary API is found in RandomGenerator and RandomGeneratorFactory. Further description can be found in the JEP https://openjdk.java.net/jeps/356 .

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (6/6 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Integration blocker
&nbsp;⚠️ The change requires a CSR request to be approved.

### Issue
 * [JDK-8248862](https://bugs.openjdk.java.net/browse/JDK-8248862): Implement Enhanced Pseudo-Random Number Generators

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1273/head:pull/1273`
`$ git checkout pull/1273`

### CODE REVIEW

**Comment by erikj79** [comment]
It looks like you have deleted the doc directory. That can't be right?

**Comment by JimLaskey** [comment]
@erikj79  That's possible I had an intermediate doc folder before the master doc was added. I though I was careful to resolve conflict by selecting master but it might be that I picked the wrong one. Will try and correct in next round,.

**Comment by JimLaskey** [comment]
@erikj79 my local branch seems to have the right sources for doc

**Comment by JimLaskey** [comment]
@erikj79 https://github.com/JimLaskey/jdk/tree/8248862

**Comment by kevinrushforth** [comment]
> my local branch seems to have the right sources for doc

Maybe, but your branch on GitHub does not.

**Comment by magicus** [comment]
This PR looks seriously messed up:

```
JimLaskey added 30 commits on Oct 9
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators
515a5a4
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
7c25f2d
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
642022d
@JimLaskey
Merge branch 'master' into 8248862
b537e0f
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
ca90d01
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
ec2941a
@JimLaskey
Merge branch 'master' into 8248862
dbe142c
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
cca6f0f
@JimLaskey
Merge branch 'master' into 8248862
93ee16f
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
84d4704
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
b50f2ee
@JimLaskey
Merge branch 'master' into 8248862
b1d6abe
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
8afe7a2
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
2f99696
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
7c58819
@JimLaskey
Merge branch 'master' into 8248862
655ad08
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
d5860b9
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
3ac78f8
@JimLaskey
Merge branch 'master' into 8248862
e3b6e64
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
8ec7f87
@JimLaskey
Merge branch 'master' into 8248862
ed959f8
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
72195e8
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
84730c3
@JimLaskey
Merge branch 'master' into 8248862
05f0477
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
7dd81ac
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
38f534b
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
f428b61
@JimLaskey
Merge branch 'master' into 8248862
5e33872
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
0bc8e3d
@JimLaskey
Merge branch 'master' into 8248862
679f1b5
JimLaskey added 7 commits 4 days ago
@JimLaskey
8248862; Implement Enhanced Pseudo-Random Number Generators  …
17972a9
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
4f0338f
@JimLaskey
Merge branch 'master' into 8248862
7c6b9fa
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
c67c729
@JimLaskey
Merge branch 'master' into 8248862
108ea50
@JimLaskey
Merge branch 'master' into 8248862
1b31205
@JimLaskey
8248862: Implement Enhanced Pseudo-Random Number Generators  …
7469ca5
```

It might be better if you close this, rebase your actual change on top of a current master, and open a new PR instead.

The only reason this was flagged with the `build` label was due to the incorrect deletion of doc, so if you intend to go on with this PR, please remove the label after you've fixed the doc issue.

**Comment by magicus** [comment]
Ah, sorry, I see now that I expanded the commit comments that it was not just the same commit applied over and over again, as I assumed, but iterative steps from your branch while developing, that just had the same first line... sorry for the confusion about that.

**Comment by JimLaskey** [comment]
@kevinrushforth What is the recommended approach to remove the doc edit/commit?

**Comment by JimLaskey** [comment]
javadoc can be found at http://cr.openjdk.java.net/~jlaskey/prng/doc/api/java.base/java/util/random/package-summary.html

**Comment by kevinrushforth** [comment]
Presuming your master branch is in sync with the upstream jdk repo, something like the following:

```
git checkout master -- doc
git add doc
git commit -m "restore doc files"
```

File: src/java.base/share/classes/java/util/random/package-info.java [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+ * from a uniform distribution; methods for requesting values of type
+ * {@code double} chosen (pseudo)randomly from a normal distribution or from an
+ * exponential distribution; and methods for creating streams of values of type
+ * {@code int}, {@code long}, or {@code double} chosen (pseudo)randomly from a
+ * uniform distribution (such streams are spliterator-based, allowing for
+ * parallel processing of their elements). There are also static factory methods
+ * for creating an instance of a specific random number generator algorithm
+ * given its name.
+ *
+ * <p> The principal supporting class is {@link RandomGenertatorFactor}. This
```
Comment (PaulSandoz) [inline src/java.base/share/classes/java/util/random/package-info.java:50]: s/RandomGenertatorFactor/RandomGenertatorFactory

File: src/java.base/share/classes/java/util/random/package-info.java [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+ * {@link java.security.SecureRandom}.
+ *
+ * <p> For applications (such as physical simulation, machine learning, and
+ * games) that do not require a cryptographically secure algorithm, this package
+ * provides multiple implementations of interface {@link RandomGenerator} that
+ * provide trade-offs among speed, space, period, accidental correlation, and
+ * equidistribution properties.
+ *
+ * <p> For applications with no special requirements,
+ * "L64X128MixRandom" has a good balance among speed, space,
```
Comment (PaulSandoz) [inline src/java.base/share/classes/java/util/random/package-info.java:140]: The documentation assumes that the `jdk.random` module is present in the JDK image. Perhaps we need to spit the specifics to `jdk.random`?

File: src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java [hunk 3]
```diff
@@ -0,0 +1,3394 @@
...
+                    double x = (X[j] * 0x1.0p63) + ((X[j-1] - X[j]) * (double)U1);
+                    // Does the point lie below the curve?
+                    long Udiff = U2 - U1;
+                    if (Udiff < 0) {
+                        // We picked a point in the upper-right triangle.  None of those can be
+                        // accepted.  So remap the point into the lower-left triangle and try that.
+                        // In effect, we swap U1 and U2, and invert the sign of Udiff.
+                        Udiff = -Udiff;
+                        U2 = U1;
+                        U1 -= Udiff;
```
Comment (PaulSandoz) [inline src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java:1211]: Updated `U1` never used (recommend running the code through a checker e.g. use IntelliJ)

File: src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java [hunk 4]
```diff
@@ -0,0 +1,3394 @@
...
+            result[j>>2] = (result[j>>2] << 8) | seed[j];
+        }
+        // If there aren't enough seed bytes for all the words we need,
+        // use a SplitMix-style PRNG to fill in the rest.
+        int v = result[0];
+        for (int j = (m + 3) >> 2; j < n; j++) {
+            result[j] = mixMurmur32(v += SILVER_RATIO_32);
+        }
+        // Finally, we need to make sure the last z words are not all zero.
+        search: {
```
Comment (PaulSandoz) [inline src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java:331]: Nice! a rarely used feature :-)

File: src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java [hunk 5]
```diff
@@ -0,0 +1,3394 @@
...
+     * @param rng an instance of {@code RandomGenerator}, used to generate uniformly
+     *            (pseudo)randomly chosen {@code long} values
+     *
+     * @return a nonnegative {@code double} value chosen (pseudo)randomly
+     *         from an exponential distribution whose mean is 1
+     */
+    public static double computeNextExponential(RandomGenerator rng) {
+        /*
+         * The tables themselves, as well as a number of associated parameters, are
+         * defined in class java.util.DoubleZigguratTables, which is automatically
```
Comment (PaulSandoz) [inline src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java:1157]: `DoubleZigguratTables` is an inner class of `RandomSupport`

File: src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java [hunk 6]
```diff
@@ -0,0 +1,3394 @@
...
+     * nextExponential()
+     *
+     * <p> This Java class was generated automatically by a program
+     * `create_ziggurat_tables.c`.
+     *
+     * <p> Fraction of the area under the curve that lies outside the layer
+     * boxes: 0.0156 Fraction of non-box area that lies in the tail of the
+     * distribution: 0.0330
+     */
+    static class DoubleZigguratTables {
```
Comment (PaulSandoz) [inline src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java:2895]: make `final`

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+                }
+            }
+        }
+        return factoryMap;
+    }
+
+    /**
+     * Return the properties map for the specified provider.
+     *
+     * @param provider  provider to locate.
```
Comment (PaulSandoz) [inline src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:167]: Method has no such parameter

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+     *
+     * @return properties map for the specified provider.
+     */
+    @SuppressWarnings("unchecked")
+    private Map<RandomGeneratorProperty, Object> getProperties() {
+        if (properties == null) {
```
Comment (PaulSandoz) [inline src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:173]: `properties` needs to be marked volatile, and it needs to be assigned at line 182 or line 184.

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+        this.provider = provider;
+    }
+
+    /**
+     * Returns the factory map, lazily constructing map on first call.
+     *
+     * @return Map of RandomGeneratorFactory classes.
+     */
+    private static Map<String, Provider<? extends RandomGenerator>> getFactoryMap() {
+        if (factoryMap == null) {
```
Comment (PaulSandoz) [inline src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:148]: `factoryMap` needs to be marked volatile when using the double checked locking idiom.

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+                    } else if (parameterTypes.length == 1) {
+                        Class<?> argType = parameterTypes[0];
+
+                        if (argType == long.class) {
+                            ctorLong = ctorSpecific;
+                        } else if (argType == byte[].class) {
+                            ctorBytes = ctorSpecific;
+                        }
+                    }
+                }
```
Comment (PaulSandoz) [inline src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:320]: Add an `assert` statement that `ctor`, `ctorLong` and `ctorBytes` are all non-null?

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java [hunk 11]
```diff
@@ -0,0 +1,509 @@
...
+                // Do nothing
+            }
+        }
+    }
+
+    /**
+     * Ensure all the required constructors are fetched.
+     */
+    private void ensureConstructors() {
+        if (ctor == null) {
```
Comment (PaulSandoz) [inline src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:331]: This check occurs outside of the synchronized block, field may need to be marked volatile. Unsure about the other dependent fields. Might need to store values from loop in `getConstructors` in locals and then assign in appropriate order, assigning the volatile field last.

**Review by PaulSandoz — COMMENTED** [review]
I am unsure if the intent is also to support external libraries providing `RandomGenerator` implementations. Currently there is an implicit contract for properties (reflectively invoking a method returning a map with a set of entries with known keys). Since the service provider implementation is the `RandomGenerator` itself, rather than `RandomGeneratorFactory` it is harder expose the meta-data with a clearer contract.

**Comment by JimLaskey** [comment]
Need rebase

**Comment by JimLaskey** [comment]
Created new PR because of forced push: https://github.com/openjdk/jdk/pull/1292

File: src/java.base/share/classes/java/util/Random.java [hunk 12]
```diff
@@ -1221,10 +579,27 @@ private synchronized void writeObject(ObjectOutputStream s)
...
         } catch (Exception ex) { throw new Error(ex); }
     }
     private void resetSeed(long seedVal) {
         unsafe.putReferenceVolatile(this, seedOffset, new AtomicLong(seedVal));
     }
+
+    // Methods required by class AbstractSpliteratorGenerator
+
+    @Override
+    public Spliterator.OfInt makeIntsSpliterator(long index, long fence, int origin, int bound) {
```
[the remark this answers is not in the document]
Comment (JimLaskey) [reply src/java.base/share/classes/java/util/Random.java:592]: Re the properties general comment: I moved properties to RandomSupport based on the notion that the SPI work with come later.

Re makeIntsSpliterator: These methods aren't exposed in the java.util.Random API  I guess no harm done. The only solution I can think of is to create an intermediate implementor, but that leaves the methods exposed as well.

File: src/java.base/share/classes/java/util/SplittableRandom.java [hunk 13]
```diff
@@ -154,18 +150,42 @@
...
+
+    /*
+     * The equidistribution of the algorithm.
+     */
+    private static final int EQUIDISTRIBUTION = 1;
+
+    /*
+     * RandomGenerator properties.
+     */
+    static Map<RandomGeneratorProperty, Object> getProperties() {
```
[the remark this answers is not in the document]
Comment (JimLaskey) [reply src/java.base/share/classes/java/util/SplittableRandom.java:171]: Yes.

File: src/java.base/share/classes/java/util/SplittableRandom.java [hunk 14]
```diff
@@ -186,6 +206,7 @@ private SplittableRandom(long seed, long gamma) {
 
     /**
      * Computes Stafford variant 13 of 64bit mix function.
+     * http://zimbry.blogspot.com/2011/09/better-bit-mixing-improving-on.html
      */
     private static long mix64(long z) {
```
[the remark this answers is not in the document]
Comment (JimLaskey) [reply src/java.base/share/classes/java/util/SplittableRandom.java:211]: We were careful to not change the sequences (from fixed seed) generated by existing prngs. This was an edge case.

File: src/java.base/share/classes/module-info.java [hunk 15]
```diff
@@ -246,6 +247,7 @@
         jdk.jfr;
     exports jdk.internal.util.xml.impl to
         jdk.jfr;
+    exports jdk.internal.util.random;
```
[the remark this answers is not in the document]
Comment (JimLaskey) [reply src/java.base/share/classes/module-info.java:250]: I guess you are right. Until we have a defined SPI we should restrict.

[the remark this answers is not in the document]
Comment (JimLaskey) [reply src/java.base/share/classes/module-info.java:250]: On the other hand:

/Volumes/Elephant/Users/jlaskey/Git/jdk/open/src/java.base/share/classes/java/util/Random.java:77: warning: [exports] class AbstractSpliteratorGenerator in module java.base may not be visible to all clients that require this module
public class Random extends AbstractSpliteratorGenerator
                            ^
error: warnings found and -Werror specified
/Volumes/Elephant/Users/jlaskey/Git/jdk/open/src/java.base/share/classes/java/util/SplittableRandom.java:85: warning: [exports] class AbstractSplittableGenerator in module java.base may not be visible to all clients that require this module
public final class SplittableRandom extends AbstractSplittableGenerator {

File: src/jdk.random/share/classes/module-info.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+ * @use jdk.internal.util.random.RandomSupport
+ *
+ * @moduleGraph
+ * @since 16
+ */
+module jdk.random {
+    uses java.util.random.RandomGenerator;
+    uses RandomSupport;
+
+    exports jdk.random to
```
[the remark this answers is not in the document]
Comment (JimLaskey) [reply src/jdk.random/share/classes/module-info.java:53]: Removing

File: src/java.base/share/classes/java/util/random/package-info.java (hunk 1 above)
Comment (JimLaskey) [reply src/java.base/share/classes/java/util/random/package-info.java:50]: fixing

File: src/java.base/share/classes/java/util/random/package-info.java (hunk 2 above)
Comment (JimLaskey) [reply src/java.base/share/classes/java/util/random/package-info.java:140]: But jdk.random isn't really a public API.

File: src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java (hunk 3 above)
Comment (JimLaskey) [reply src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java:1211]: I noticed that before. I think it's a symmetry thing - will check with Guy.

File: src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java (hunk 5 above)
Comment (JimLaskey) [reply src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java:1157]: Late change fixing.

File: src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java (hunk 6 above)
Comment (JimLaskey) [reply src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java:2895]: fixing

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java (hunk 7 above)
Comment (JimLaskey) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:167]: Fixing

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java (hunk 8 above)
Comment (JimLaskey) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:173]: One of them foggy days.

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java (hunk 9 above)
Comment (JimLaskey) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:148]: fixing

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java (hunk 10 above)
Comment (JimLaskey) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:320]: Only `ctor` is required but yes.

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java (hunk 11 above)
Comment (JimLaskey) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:331]: okay

**Comment by forax** [comment]
LGTM

**Pull request closed without merging**